File: src/content/docs/topics/cs/coding-problems/binary-search/074-search-a-2d-matrix-approach2.go
package main

import "fmt"

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func searchMatrix(matrix [][]int, target int) bool {
	for _, row := range matrix {
		n := len(row)
		if row[0] <= target && target <= row[n-1] {
			lo, hi := 0, n-1
			for lo <= hi {
				mid := (lo + hi) / 2
				if row[mid] == target {
					return true
				}
				if row[mid] < target {
					lo = mid + 1
				} else {
					hi = mid - 1
				}
			}
		}
	}
	return false
}

func runTests() {
	m := [][]int{{1, 3, 5, 7}, {10, 11, 16, 20}, {23, 30, 34, 60}}
	assert(searchMatrix(m, 3) == true)
	assert(searchMatrix(m, 13) == false)
	assert(searchMatrix([][]int{{1}}, 1) == true)
	assert(searchMatrix([][]int{{1}}, 2) == false)
	assert(searchMatrix([][]int{{1, 3}}, 3) == true)
	assert(searchMatrix([][]int{{1}, {3}}, 1) == true)
	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/binary-search/074-search-a-2d-matrix-approach2.py
from bisect import bisect_left

def search_matrix(matrix: list[list[int]], target: int) -> bool:
    for row in matrix:
        if row[0] <= target <= row[-1]:
            i = bisect_left(row, target)
            if i < len(row) and row[i] == target:
                return True
    return False

m = [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]]
assert search_matrix(m, 3) == True
assert search_matrix(m, 13) == False
assert search_matrix([[1]], 1) == True
assert search_matrix([[1]], 2) == False
assert search_matrix([[1, 3]], 3) == True
assert search_matrix([[1], [3]], 1) == True
print("all tests pass")
